### PR TITLE
change dump format to include raw_data

### PR DESCRIPTION
**Pull request by willr3**
This changes the `--dump` json format to include the `raw_data` in `data` along with the `metric_name` and `query` that produced the `raw_data`. This also moves the `jenkins_env` and `netobserv_env` objects outside the `RESULST['data']` array to simplify processing when converting to `clean_data` for elastic search upload.

The goal of this change is to persist the data from the System Under Test (SUT) in a format that closely matches the format from the underlying tools (e.g. prometheus). This format will help support other tools' use of the data and decrease the json document size by decreasing the number of copies of the `metric` prometheus response field that is currently persisted in the `metadata` field. 

Note: Please check the PR as though it was written by a 5-year-old. I do not have an environment to test the python code and about half the time I had to correct myself from trying to write in javascript notation (.e.g `RESULTS.data` instead of `RESULTS['data']`)

### CODE REVIEW

File: scripts/nope.py [hunk 1]
```diff
@@ -304,6 +304,23 @@ def dump_data_locally(timestamp, partial=False):
     # return if no issues
     return None
 
+def format_for_upload():
+    ''' formats the data in RESULTS into a de-normalized payload for elasticsearch
+        returns a map with the de-normalized array under the 'data' key
+    '''
+    payload = {"data":[]}
```
Comment (nathan-weinberg) [inline scripts/nope.py:311]: I'm thinking we might not need to use the `data` var at all here but could just initialize the `payload` as an empty list, see my other comment below

File: scripts/nope.py [hunk 2]
```diff
@@ -304,6 +304,23 @@ def dump_data_locally(timestamp, partial=False):
     # return if no issues
     return None
 
+def format_for_upload():
+    ''' formats the data in RESULTS into a de-normalized payload for elasticsearch
+        returns a map with the de-normalized array under the 'data' key
+    '''
```
Comment (nathan-weinberg) [inline scripts/nope.py:310]:
```suggestion
    ''' formats the data in RESULTS into a de-normalized payload for elasticsearch
        returns a dict with the de-normalized list under the 'data' key
    '''
```
JS/Python terms :wink: but if we go the approach I suggest below we could just do
```suggestion
    ''' formats the data in RESULTS into a de-normalized list for elasticsearch upload compatability
    '''
```

File: scripts/nope.py [hunk 3]
```diff
@@ -485,18 +504,17 @@ def main():
...
+    RESULTS["netobserv_env"] = get_netobserv_env_info()
 
     # get prometheus data
     for entry in QUERIES:
         metric_name = entry['metricName']
         query = entry['query']
         raw_data = run_query(metric_name, query)
-        clean_data = process_query(metric_name, query, raw_data)
-        RESULTS["data"].extend(clean_data)
+        RESULTS["data"].append({"raw_data": raw_data, "query" : query, "metric_name" : metric_name})
```
Comment (nathan-weinberg) [inline scripts/nope.py:517]:
```suggestion
        RESULTS["prometheus_data"].append({"raw_data": raw_data, "query" : query, "metric_name" : metric_name})
```

File: scripts/nope.py [hunk 4]
```diff
@@ -317,8 +334,10 @@ def upload_data_to_elasticsearch():
         retry_on_timeout=True
     )
 
+    formatted = format_for_upload()
+
     start = time.time()
-    for item in RESULTS['data']:
+    for item in formatted['data']:
```
Comment (nathan-weinberg) [inline scripts/nope.py:340]:
```suggestion
    for item in documents:
```

File: scripts/nope.py [hunk 5]
```diff
@@ -317,8 +334,10 @@ def upload_data_to_elasticsearch():
         retry_on_timeout=True
     )
 
+    formatted = format_for_upload()
```
Comment (nathan-weinberg) [inline scripts/nope.py:337]:
```suggestion
    documents = format_data_for_upload()
```

File: scripts/nope.py [hunk 6]
```diff
@@ -304,6 +304,23 @@ def dump_data_locally(timestamp, partial=False):
     # return if no issues
     return None
 
+def format_for_upload():
```
Comment (nathan-weinberg) [inline scripts/nope.py:307]:
```suggestion
def format_data_for_upload():
```

File: scripts/nope.py [hunk 7]
```diff
@@ -304,6 +304,23 @@ def dump_data_locally(timestamp, partial=False):
...
+
+    if 'jenkins_env' in RESULTS:
+        payload["data"].append(RESULTS['jenkins_env'])
+    if 'netobserve_env' in RESULTS:
+        payload["data"].append(RESULTS['netobserve_env'])
+
+    if 'data' in RESULTS and len(RESULTS['data']) > 0:
+        for item in RESULTS['data']:
+            clean_data = process_query(item['metric_name'], item['query'], item['raw_data'])
+            payload["data"].extend(clean_data)
```
Comment (nathan-weinberg) [inline scripts/nope.py:321]:
```suggestion
    if 'jenkins_env' in RESULTS:
        payload.append(RESULTS['jenkins_env'])
    if 'netobserv_env' in RESULTS:
        payload.append(RESULTS['netobserv_env'])
    if 'prometheus_data' in RESULTS and len(RESULTS['prometheus_data']) > 0:
        for item in RESULTS['prometheus_data']:
            clean_data = process_query(item['metric_name'], item['query'], item['raw_data'])
            payload["prometheus_data"].extend(clean_data)
```

File: scripts/nope.py (hunk 1 above)
Comment (willr3) [reply scripts/nope.py:311]: I agree, no need to complicate the internal representation. I commited the recommended changes. Do you want me to squash them back into a single commit and rebase?

Comment (nathan-weinberg) [reply scripts/nope.py]: @willr3 Nah no need for backwards compatibility with the old format, I will retest the changes today if I can, early next week at the latest - appreciate the updates!

File: scripts/nope.py [hunk 8]
```diff
@@ -304,6 +304,22 @@ def dump_data_locally(timestamp, partial=False):
...
+    '''
+    payload = []
+
+    if 'jenkins_env' in RESULTS:
+        payload.append(RESULTS['jenkins_env'])
+    if 'netobserv_env' in RESULTS:
+        payload.append(RESULTS['netobserv_env'])
+    if 'prometheus_data' in RESULTS and len(RESULTS['prometheus_data']) > 0:
+        for item in RESULTS['prometheus_data']:
+            clean_data = process_query(item['metric_name'], item['query'], item['raw_data'])
```
Comment (nathan-weinberg) [inline scripts/nope.py:319]:
```suggestion
            clean_data = process_query(
                uuid,
                item['metric_name'],
                item['query'],
                item['raw_data']
            )
```

File: scripts/nope.py [hunk 9]
```diff
@@ -304,6 +304,22 @@ def dump_data_locally(timestamp, partial=False):
...
+def format_data_for_upload():
+    ''' formats the data in RESULTS into a de-normalized payload for elasticsearch
+        returns a de-normalized list
+    '''
+    payload = []
+
+    if 'jenkins_env' in RESULTS:
+        payload.append(RESULTS['jenkins_env'])
+    if 'netobserv_env' in RESULTS:
+        payload.append(RESULTS['netobserv_env'])
```
Comment (nathan-weinberg) [inline scripts/nope.py:316]:
```suggestion
    if 'jenkins_env' in RESULTS:
        payload.append(RESULTS['jenkins_env'])
        uuid = RESULTS['jenkins_env]['uuid']
    if 'netobserv_env' in RESULTS:
        payload.append(RESULTS['netobserv_env'])
        if uuid is None:
            uuid =  RESULTS['netobserv_env']['uuid']
        else:
            assert uuid == RESULTS['netobserv_env']['uuid'], "UUIDs for jenkinsEnv and netobservEnv did not match, data may be malformed"
```

File: scripts/nope.py [hunk 10]
```diff
@@ -304,6 +304,22 @@ def dump_data_locally(timestamp, partial=False):
     # return if no issues
     return None
 
+def format_data_for_upload():
+    ''' formats the data in RESULTS into a de-normalized payload for elasticsearch
+        returns a de-normalized list
+    '''
+    payload = []
```
Comment (nathan-weinberg) [inline scripts/nope.py:311]:
```suggestion
    uuid = None
    payload = []
```

File: scripts/nope.py [hunk 11]
```diff
@@ -317,8 +333,10 @@ def upload_data_to_elasticsearch():
         retry_on_timeout=True
     )
 
+    documents = format_data_for_upload()
+
```
Comment (nathan-weinberg) [inline scripts/nope.py:337]: Can remove extra space here

**Review by nathan-weinberg — CHANGES_REQUESTED** [review]
@willr3 I found one bug with ES uploading where UUID isn't be recorded for the datapoints - my suggestions should fix it but if you notice anything amiss with the logic lmk

**Comment by nathan-weinberg** [comment]
Following changes would be needed to `process_query` as well (basically just add new `uuid` param)
```python
def process_query(uuid, metric_name, query, raw_data):
    ''' takes in a UUID, Prometheus metric name, query and its successful execution result
        reformats data into format desired for Elasticsearch upload
        returns dict object with post-processed data
    '''

    # intialize post-processing data object
    clean_data = []

    # capture metadata and value artifacts from prometheus return object
    try:
        for result in raw_data["data"]["result"]:
            for data_point in result["values"]:
                metadata = result["metric"]
                metadata["query"] = query
                iso_timestamp = get_iso_timestamp(data_point[0])
                clean_data.append(
                    {
                        "uuid": uuid,
                        "metric_name": metric_name,
                        "data_type": "datapoint",
                        "iso_timestamp": iso_timestamp,
                        "unix_timestamp": data_point[0],
                        "value": float(data_point[1]),
                        "metadata": metadata
                    }
                )
```

**Comment by willr3** [comment]
Excellent catch on the `uuid` I updated the PR with the fix to `process_query` and the spacing comment as well as the other discussions you raised. 
Do we want `NOPE` to be able to work with older dump formats?

File: scripts/nope.py (hunk 11 above)
Comment (willr3) [reply scripts/nope.py]: I removed the extra line in the newest update

**Comment by nathan-weinberg** [comment]
> Excellent catch on the `uuid` I updated the PR with the fix to `process_query` and the spacing comment as well as the other discussions you raised. Do we want `NOPE` to be able to work with older dump formats?

Nah don't think we'll need that - I did find one issue with the code changes that needs to be fixed, see my comment above

**Review by nathan-weinberg — APPROVED** [review]
Full test run: https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/scale-ci/job/nweinber-e2e-benchmarking-multibranch-pipeline/job/netobserv-dump-raw/15/

Things are looking good to me, want my colleagues to review before merging but as it stands I'm happy with this on my side.

**Review by memodi — APPROVED** [review]
code changes look good to me.

**Pull request merged**